scalding-macros/src/main/scala/com/twitter/scalding/macros/impl/ordered_serialization/TreeOrderedBuf.scala

```diff
@@ -0,0 +1,299 @@
+/*
+ Copyright 2014 Twitter, Inc.
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+ http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ */
+package com.twitter.scalding.macros.impl.ordered_serialization
+
+import com.twitter.scalding._
+import com.twitter.scalding.serialization.OrderedSerialization
+import com.twitter.scalding.serialization.JavaStreamEnrichments
+import java.io.InputStream
+import scala.reflect.macros.Context
+import scala.language.experimental.macros
+import scala.util.control.NonFatal
+
+object CommonCompareBinary {
+  import com.twitter.scalding.serialization.JavaStreamEnrichments._
+
+  // If the lengths are equal and greater than this number
+  // we will compare on all the containing bytes
+  val minSizeForFulBinaryCompare = 24
+
+  /**
+   * This method will compare two InputStreams of given lengths
+   * If the inputsteam supports mark/reset (such as those backed by Array[Byte]),
+   * and the lengths are equal and longer than minSizeForFulBinaryCompare we first
+   * check if they are byte-for-byte identical, which is a cheap way to avoid doing
+   * potentially complex logic in binary comparators
+   */
+  final def earlyEqual(inputStreamA: InputStream,
+    lenA: Int,
+    inputStreamB: InputStream,
+    lenB: Int): Boolean =
+    (lenA > minSizeForFulBinaryCompare &&
+      (lenA == lenB) &&
+      inputStreamA.markSupported &&
+      inputStreamB.markSupported) && {
+        inputStreamA.mark(lenA)
+        inputStreamB.mark(lenB)
+
+        var pos: Int = 0
+        while (pos < lenA) {
+          val a = inputStreamA.read
```

Comment:
are these streams usually backed by a byte array? If not, reading 1 byte at a time vs. chunks of bytes could matter, though maybe BufferedInputStreams handle that case for us too?